tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/MetaAttributeGenerationVisitor.java

```diff
@@ -138,14 +138,20 @@ private AnnotationMetaAttribute createMetaCollectionAttribute(DeclaredType decla
 				accessTypeInfo.setDefaultAccessType( entity.getEntityAccessTypeInfo().getAccessType() );
 			}
 		}
-		if ( collection.equals( "javax.persistence.metamodel.MapAttribute" ) ) {
+		if ( collection.equals( Constants.MAP_ATTRIBUTE ) ) {
 			return createAnnotationMetaAttributeForMap( declaredType, element, collection, targetEntity );
 		}
-		else {
-			return new AnnotationMetaCollection(
-					entity, element, collection, getElementType( declaredType, targetEntity )
+		if ( TypeUtils.containsAnnotation( element, Constants.CONVERT ) ||
+				TypeUtils.containsAnnotation( element, Constants.HIBERNATE_TYPE )	) {
+			return new AnnotationMetaSingleAttribute(
+					entity,
+					element,
+					TypeUtils.toTypeString( declaredType )
```

Comment:
The only downside to using `TypeUtils.toTypeString( declaredType )` is that the static definition in the static model class will be generated with fully qualified class names rather than using an import clause, effectively looking like this:
````
public static volatile SingularAttribute<SomeEntity, List<java.lang.String>> someVariable;
````
I think what we can do is address this by adjusting `ImportContextImpl#importType` as follows:
````
String additionalTypePart = null;
if ( fqcn.indexOf( '<' ) >= 0 ) {
	additionalTypePart = result.substring( fqcn.indexOf( '<' ) );
	// start new functionality to recursively handle importing generic types
	String genericTypes = additionalTypePart.substring( 1, additionalTypePart.length() - 1 );
	additionalTypePart = "<" + importType( genericTypes ) + ">";
	// end new functionality
	result = result.substring( 0, fqcn.indexOf( '<' ) );
	fqcn = result;
}
else if ( fqcn.indexOf( '[' ) >= 0 ) {
	additionalTypePart = result.substring( fqcn.indexOf( '[' ) );
	result = result.substring( 0, fqcn.indexOf( '[' ) );
	fqcn = result;
}
````
We'd need to check if this does not introduce any regressions, particularly around use cases where the imported types resulted in type clashes.  I believe the subsequent part of this method handle this, but would just want to verify.  At least to me, this ends up looking like cleaner class code rather than inline fully qualified class names.

Reply:
If we do this change, we get the following error:

> error: ';' expected
> import java.lang.String,java.lang.String;
>                        ^

It looks like this logic is further used for imports.